### DISCUSSION TITLE
How is copilot X so much less capable than ChatGPT?

### DISCUSSION THREAD

**Discussion by MattesR** [community/community#60355] [Copilot Conversations] [answered]
### Select Topic Area

Product Feedback

### Body

At first I was pretty excited about Copilot X, because I do lots of Data Science and Data Wrangling related task, where describing a problem or an outcome often is pretty easy, but still after a few years of coding, it may be troublesome (or sometimes even hard) to nail it in pandas, which has a sometimes convoluted syntax.
I give you an example for a task that I wanted to have just done and for which I thought Copilot X _should be _ perfect.

> from a list of dataframes with same index and same columns, create a dataframe that also has the same index and columns and in each cell has the standard variation and mean of the dataframes in the list of dataframes

this is copilots code:

```python
import pandas as pd

# Example list of DataFrames
dfs = [
    pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]}),
    pd.DataFrame({'A': [4, 5, 6], 'B': [7, 8, 9]}),
    pd.DataFrame({'A': [7, 8, 9], 'B': [10, 11, 12]})
]

# Concatenate the DataFrames into a single DataFrame
df = pd.concat(dfs, axis=0, keys=range(len(dfs)))

# Calculate the mean and standard deviation of each cell value in the DataFrame
df_stats = df.apply(lambda x: pd.Series({'mean': x.mean(), 'std': x.std()}))

# Print the resulting DataFrame
print(df_stats)
```

The code from Copilot doesn't work on this, it also has misunderstood the assignment, as gives the mean of the columns A and B instead of the mean of the cells. It also made other suggestions that didn't work.

I have 2 issues with that:

1. Right now Copilot X seems to have basically no memory of the stuff it generated before, which makes "arguing" with it about mistakes in its code very bothersome.
2. it just is so much worse than ChatGPT, which nailed the assignment with exactly the same input. 
That leads me to the question: Why bother with Copilot X when ChatGPT is more helpful right now? Maybe that makes sense from a business perpective (I assume ChatGPT is more ressource-hungry) but from a consumer perspective, why would I pay 10 bucks for something, that is inferior to something that I can (currently) use for free?

**Comment by GhostOf0days** [comment]
I wonder if there's a way Github could use a Hopfield network for memory or if OpenAI alone can do it.

**Comment by ronakjkothari** [comment]
While there are some cases in which CopilotX may be unmatched in comparison to ChatGPT, in most cases, CopilotX will write cleaner and more effective code than ChatGPT. Unlike ChatGPT, CopilotX was trained to write code for even niche cases. ChatGPT can recite the code it will take to write common algorithms. With my experience with CopilotX, it can read already written code in a project and write code accordingly, using the previously declared variables and such.

Furthermore, CopilotX will improve in the programming realm, whereas ChatGPT will be more broad in scope, and may not be able to keep up in the coding realm.

However, I will give ChatGPT this: it can create relatively decent code for common algorithms, and can sometimes beat CopilotX in certain cases.

Perhaps it was different when I tried it, but CopilotX was able to remember previous information.

**Reply by MattesR** [reply, marked as answer]
First of all, I don't doubt that Copilot will improve, mostly because Microsoft has the capability to improve it and because it really has to in order to live up to the expectations that Github itself is generating about it.
Whether the code of Copilot is cleaner and more efficient may depend on the programming language that is used. For AI and Data Science/Big Data stuff in Python, Copilot is pretty hit or miss. It can generate code that fits the overall code/intent in a file, but it also gets stuck on writing comments or checks unnecessary cases one after another without actually touching the assignment. However, that all isn't even the function which I want to compare, but the chat client that was added with Copilot X. 
My OP gave a good example to what I expect. Posing a question/instruction and getting functional code that solve a task. ChatGPT can do that, Copilot X doesn't. Further, ChatGPT let's you give additional feedback, request changes to the code etc; copilot X doesn't. 
To me the question arises, why I pay ten bucks for Copilot (X), when I can just ask ChatGPT and get more capable answers. Sure, it's somewhat more of a hassle to ask ChatGPT, but the most important feature of a code generator is working code, and if Copilot X fails at delivering it for the cases that I need it, I question it's usefulness.

I feel like Github is trying to find out with how small models they can run Copilot X. This model is definitely to small, it needs to get beefier in order to be more useful.

**Comment by denisdasilvarocha** [comment]
Because its the same old Codex model and not GPT-4, not worth buying Copilot atm imo
